[PERF] all build jobs used for performance testing are now possible to run inividually

e-kharion opened the pull request:
For perfBisect we need to have a possibility to build specific builds on demand.

This PR should not change the behavior by default, it only adds possibility to build only selected configurations.

caaavik-msft commented on eng/pipelines/performance/templates/perf-arm64-build-jobs.yml:
```diff
@@ -12,7 +14,7 @@ jobs:
         runtimeFlavor: mono
         platforms:
-        - linux_arm64
+        - ${{ if eq(parameters.linux, true) }}: linux_arm64
```
```suggestion
        - ${{ if eq(parameters.linux, true) }}:
          - linux_arm64
```
Convention is typically to put the list entries on a separate line, I believe this is the correct syntax for it

caaavik-msft commented on eng/pipelines/performance/templates/perf-arm64-build-jobs.yml:
```diff
@@ -12,7 +14,7 @@ jobs:
         runtimeFlavor: mono
         platforms:
-        - linux_arm64
+        - ${{ if eq(parameters.linux, true) }}: linux_arm64
```
When doing this, if `linux` is false does it cause no build job to created at all? I expect it should but just making sure

e-kharion replied:
good point, originally I was not paying much attention to checking validity of configuration, but I improved the conditions to cover also those cases

caaavik-msft commented on eng/pipelines/performance/templates/perf-build-jobs.yml:
```diff
@@ -1,143 +1,164 @@
-        nameSuffix: AOT
-        isOfficialBuild: false
-        postBuildSteps:
+  - ${{ if or(containsValue(parameters.jobsAndPlatforms, 'coreclr:linux_x64'), containsValue(parameters.jobsAndPlatforms, 'coreclr:windows_x64'), containsValue(parameters.jobsAndPlatforms, 'coreclr:windows_x86'), containsValue(parameters.jobsAndPlatforms, 'coreclr:linux_musl_x64')) }}:
```
Is this if statement needed? If all of these are false, then the platforms list will be false, and then it should be equivalent.

e-kharion replied:
I'd rather keep it there. There are defined postBuildSteps and I believe by keeping it there we should make it more error prone

caaavik-msft commented on eng/pipelines/performance/templates/perf-build-jobs.yml:
```diff
@@ -1,143 +1,164 @@
 parameters:
   perfBranch: 'main' # Currently only used for building the BDN App
+  jobsAndPlatforms: [
```
Rather than needing this complicated jobsAndPlatforms object, could we split out this `perf-build-jobs.yml` into multiple simpler yaml files? We can still keep this file as the root one and all it does is call each of the other simpler yaml files. Then we don't need to implement as many parameters, the caller can instead just pick and choose which of the simpler yaml files to run with a reduced parameter set.

I am thinking we could do `perf-mono-build-jobs.yml`, `perf-coreclr-build-jobs.yml`, `perf-bdn-build-jobs.yml`

e-kharion replied:
reworked

caaavik-msft commented on eng/pipelines/performance/templates/perf-ios-scenarios-build-jobs.yml:
```diff
@@ -1,51 +1,55 @@
 parameters:
   hybridGlobalization: true
+  buildMono: true
+  buildNative: true
```
I think let's call these two params `mono` and `nativeAot`.

matouskozak commented on eng/pipelines/performance/templates/perf-build-jobs.yml:
```diff
@@ -1,143 +1,164 @@
   perfBranch: 'main' # Currently only used for building the BDN App
+  jobsAndPlatforms: [
+    'coreclr:linux_x64', 'coreclr:windows_x64', 'coreclr:windows_x86', 'coreclr:linux_musl_x64',
+    'monoAot',
+    'monoAndroid',
+    'mono',
+    'monoPacks',
+    'perfBDNApp'
```
What all of these represent? 

On the Mono desktop side, we have:
- MonoAOT x64/arm64
- MonoInterpreter x64/arm64
- MonoJIT x64/arm64
- WasmAOT
- WasmInterpreter

On Mobile, there is a bunch more:
- Android runtime sample
- Android MAUI sample
- iOS Mono/NativeAOT samples
- iOS MAUI Mono/NativeAOT samples

Maybe more? I just want to make sure the listed entries cover all scenarios.

e-kharion replied:
I welcome any idea on renaming if it's confusing. It should cover every build "configuration" we do for performance testing, however.
If there is any important configuration missing, we should address it.

e-kharion replied:
I split up the file based on Cameron's suggestion

matouskozak commented on eng/pipelines/performance/templates/perf-build-jobs.yml:
```diff
@@ -1,143 +1,164 @@
-          vmImage: 'macos-13'
-        postBuildSteps:
-          - template: /eng/pipelines/performance/templates/build-perf-bdn-app.yml
+  - ${{ if containsValue(parameters.jobsAndPlatforms, 'monoAndroid') }}:
```
fyi: this will need to align with https://github.com/dotnet/runtime/pull/112939

matouskozak commented on eng/pipelines/performance/templates/perf-build-jobs.yml:
```diff
@@ -1,143 +1,164 @@
+              artifactName: BuildArtifacts_$(osGroup)$(osSubgroup)_$(archType)_$(_BuildConfig)_mono
+              displayName: Build Assets
+              
+  - ${{ if containsValue(parameters.jobsAndPlatforms, 'monoPacks') }}:
+    # build mono runtime packs
+    - template: /eng/pipelines/common/platform-matrix.yml
+      parameters:
+        jobTemplate: /eng/pipelines/common/global-build-job.yml
+        buildConfig: release
+        runtimeFlavor: mono
+        platforms:
+        - android_arm64
+        jobParameters:
+          buildArgs: -s mono+libs+host+packs -c $(_BuildConfig)
+          nameSuffix: Mono_Packs
+          isOfficialBuild: false
+          postBuildSteps:
+            - template: /eng/pipelines/common/upload-intermediate-artifacts-step.yml
+              parameters:
+                name: MonoRuntimePacks
+                isOfficialBuild: false
```
Not related to this PR as it is just a copy from the old code, but why do we build mono runtime packs with `android_arm64` platform?

DrewScoggins replied:
I also do not know. I assume they were getting used at some time. I suppose that is no longer the case?

LoopedBard3 replied:
The mono runtime packs are used in the building of the perfBDN app. As to why android_arm64 over something like windows_arm64, I don't recall anything specific. I think it was to ensure it was labeled android in the pipeline view for easier picking out, but I don't think there is a requirement for android_arm64.

LoopedBard3 replied:
Actually, I took another look, the other reason that android_arm64 was chosen is that influences the arch and os that the build.sh command passes in. So by choosing android_arm64, the android packages and what not are built targeting the android OS.

Commit:
all build jobs used for performance testing are now possible to run individually
added possibility to configure maui framework to use
perf-build-jobs splitted to multiple templates
perf-arm64-builds moved to a different yml and yml files reorganized

DrewScoggins commented:
Do we also plan to add these configs as selectable values when starting a manual build so that we can limit what we are building for testing scenarios?

e-kharion commented:
> Do we also plan to add these configs as selectable values when starting a manual build so that we can limit what we are building for testing scenarios?

It should be easy if there is buy-in for it. I'd add it, however, in a separate PR, so this one can be merged to unblock follow-up work.
